Complete trade protocol

chimp1984 opened the pull request:
Depends on https://github.com/sqrrm/bisq-desktop/pull/14

chimp1984 committed:
Add common interface for tradePeer.
Move TradingPeer to trade package.

Only stub now...

chimp1984 committed:
Remove payment method handling

chimp1984 committed:
Add generics for TradePeer

chimp1984 committed:
Rename getTradingPeer to getTradePeer

chimp1984 committed:
Replace TradingPeer BsqSwapTradePeer in BsqSwapProtocolModel

chimp1984 committed:
Add the 4 variants to trade and protocols

Comment out all existing code in protocol domain. Will get re-added step by step...

chimp1984 committed:
Trade protocol update according to other trade protocol

- Follow the buyer/seller and maker/taker perspectives
  according to the domain context
- Use the protocol model only for state data required as
  shared model for the tasks
- Add a tradePeer model for holding the peers data
- Keep logic in tasks and use a util class for re-used methods

chimp1984 committed:
Use tx confidence listener in SellerSetupTxListener

chimp1984 committed:
Cleanups

chimp1984 committed:
Rename bsqSwapProtocolModel to protocolModel

chimp1984 committed:
Rename bsqSwapTrade to trade

chimp1984 committed:
Rename TradeProtocolModel to ProtocolModel

chimp1984 committed:
Rename bsqSwapTrade to trade

chimp1984 committed:
Move package bisq.core.trade.messages to bisq.core.trade.protocol.messages

As messages are part of the protocol its better inside the protocol package

chimp1984 committed:
Move ClosedTradableManager and FailedTradesManager inside bisq.core.trade.model

chimp1984 committed:
Remove references from multisigbased trade domain from generic trade classes
Use interfaces for state and phase

chimp1984 committed:
Use concrete state/phase in impl classes
Cleanups

chimp1984 committed:
Move contract into trade package

chimp1984 committed:
Use Trade as type in tradableList
Make delegates to bsqSwapTradeManager more explicit
Cleanup stream concats
Add BsqTradeDetailsWindow instead using OfferDetailsWindow
(not impl. changes yet in window)

chimp1984 committed:
Remove offer only once trade is completed (tx visisble if maker is seller)
Pass bsqWalletService to getTransaction as provider is not set for
closed trades but we want to access it in views.

chimp1984 committed:
Rename classes

chimp1984 committed:
Rename trade package to bisq_v1

chimp1984 committed:
Move TradeProtocolFactory level up

chimp1984 committed:
Rename bsqswap package to bsq_swap

chimp1984 committed:
Start work on BsqTradeDetailsWindow

chimp1984 committed:
Add comment to make fee allocation in tx more clear

chimp1984 committed:
Rename BsqSwapWalletWatcher to BsqSwapOfferManager
Fix params in test

chimp1984 committed:
Add methods to DaoFacade and DaoStateService

chimp1984 committed:
Add fee listener

chimp1984 committed:
Use Singleton annotation

chimp1984 committed:
Let BsqSwapAccountPayload extend PaymentAccountPayload
We do not use an address but use that account only as
proxy for marking Bsq Swaps

chimp1984 committed:
Add comment

chimp1984 committed:
Increase fee tolerance.
Fee can be quite volatile so we allow larger deviations

chimp1984 committed:
Rename getTradeVolume to getVolume

chimp1984 committed:
Add getBsqTradeAmount to BsqSwapCalculation
Use that method in BsqTrade.getBsqTradeAmount()

chimp1984 committed:
Check for witness before setting it

chimp1984 committed:
Remove unnecessary check, add comment

chimp1984 committed:
Add todo

chimp1984 committed:
Add comment

chimp1984 committed:
Refactor tx creation and verification

Add message with finalized tx to be sent to seller.
Add Publish tx task to sellers side (only dont if
tx not already received)

chimp1984 committed:
Add handling for trade complete in case we have not got it set already

chimp1984 committed:
Extract super class for sending messages

chimp1984 committed:
Cleanups

sqrrm commented on core/src/main/java/bisq/core/offer/BsqSwapOfferManager.java:
```diff
@@ -41,6 +43,17 @@
 import lombok.extern.slf4j.Slf4j;
 
+/**
+ * Listen on wallet change events, BSQ offer updates and feeService updates and
+ * recalculates the required funds for the open BSQ offers. If funds in wallets are not
+ * sufficient we close offers recursively until funds are sufficient.
+ * We notify listeners when we de-active an offer, so the UI can show a popup to the user.
+ * We do not re-activate offers after sufficient funds are available again but leave that
+ * to the user.
+ * At the moment there is no edit feature for BSQ offers so we do not need to treat the
+ * case that existing offers change, but once that is implemented that need to be covered
+ * as well.
```
This comment makes it sound like the sum of all open offers must be funded, but I think that's wrong. Should be enough to have each offer funded individually. Once an offer is taken the funding calculation will be done again.

I don't think it makes sense to allow editing offers that have to trade fee tx. Better to just remove the offer and create a new one. This process could be made easy by copying the details from the original offer and filling in the create offer with those details. Edited offers can cause trouble when taker thinks they're taking the offer with the old details but maker have already updated.

chimp1984 replied:
Yes, true. 

Edited offers are new offers just with same offerId. But here we could create it completely new. But not planned anyway atm to support that.

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/BsqSwapCalculation.java:
```diff
@@ -129,7 +104,7 @@ public static Coin getBsqTradeAmount(Volume volume) {
 
-    private static long getTradeFeeConvertedToBtc(BsqSwapTrade bsqSwapTrade, long tradeFee) {
-        return bsqSwapTrade.getPrice().getValue() * tradeFee / 100;
+    private static long getTxFee(BsqSwapTrade trade, int txSize, long tradeFee) {
```
Name could be improved since this is mining fee - sats burnt as trading fee. It could be negative for high value swaps with a lot of BSQ burnt. Intuitively a fee should be positive. Maybe something like `miningFeeAdjustment`

chimp1984 replied:
Agree... will rename

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/BsqSwapCalculation.java:
```diff
@@ -42,75 +45,47 @@
 
+    // Seller
+    public static Coin getSellersBtcInputValue(BsqSwapTrade trade, int sellersTxSize, long sellersTradeFee) {
+
```
Empty line

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/tasks/seller/ProcessTxInputsMessage.java:
```diff
@@ -45,41 +47,53 @@ public ProcessTxInputsMessage(TaskRunner<BsqSwapTrade> taskHandler, BsqSwapTrade
+            checkArgument(inputs.size() == numValidBsqInputs,
+                    "Some of the buyers BSQ inputs are not from spendable BSQ utxos according to our DAO state data.");
+
+            long change = message.getBsqChange();
```
Why trust the peer change amount rather than calculating it from BSQ inputs and BSQ payout? I don't see this value being verified later on.

chimp1984 replied:
Ok, will add that

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/tasks/seller/ProcessTxInputsMessage.java:
```diff
@@ -45,41 +47,53 @@ public ProcessTxInputsMessage(TaskRunner<BsqSwapTrade> taskHandler, BsqSwapTrade
-            String buyersBtcAPayoutAddress = data.getBuyersBtcPayoutAddress();
-            checkNotNull(buyersBtcAPayoutAddress, "buyersBtcAPayoutAddress must not be null");
-            checkArgument(!buyersBtcAPayoutAddress.isEmpty(), "buyersBtcAPayoutAddress must not be empty");
+            String buyersBtcPayoutAddress = message.getBuyersBtcPayoutAddress();
+            checkNotNull(buyersBtcPayoutAddress, "buyersBtcPayoutAddress must not be null");
+            checkArgument(!buyersBtcPayoutAddress.isEmpty(), "buyersBtcPayoutAddress must not be empty");
```
Would be good to check that the address is a valid address

chimp1984 replied:
Hm... if the peer send an invalid address, the tx would fail. If its just a wrong address its the problem of the peer. Not sure if it really adds any security here.

sqrrm replied:
It's more that it would be in line with the rest of the verification. Better to fail early than failing on signing the tx at a later step.

chimp1984 replied:
Added it

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/tasks/seller/ProcessTxInputsMessage.java:
```diff
@@ -45,41 +47,53 @@ public ProcessTxInputsMessage(TaskRunner<BsqSwapTrade> taskHandler, BsqSwapTrade
-            String buyersBsqChangeAddress = data.getBuyersBsqChangeAddress();
+            String buyersBsqChangeAddress = message.getBuyersBsqChangeAddress();
             checkNotNull(buyersBsqChangeAddress, "buyersBsqChangeAddress must not be null");
             checkArgument(!buyersBsqChangeAddress.isEmpty(), "buyersBsqChangeAddress must not be empty");
```
Check it's a valid address

chimp1984 replied:
Same as above

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/tasks/seller/ProcessTxInputsMessage.java:
```diff
@@ -45,41 +47,53 @@ public ProcessTxInputsMessage(TaskRunner<BsqSwapTrade> taskHandler, BsqSwapTrade
-            tradePeer.setBtcAddress(data.getBuyersBtcPayoutAddress());
-            tradePeer.setBsqAddress(data.getBuyersBsqChangeAddress());
+            tradePeer.setInputs(inputs);
+            tradePeer.setChange(change);
```
Setting the change value from message might not be the safest.

chimp1984 replied:
Will add verification

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/tasks/seller/SellerCreatesAndSignsTx.java:
```diff
@@ -56,60 +56,54 @@ protected void run() {
+            String sellersBsqPayoutAddress = protocolModel.getBsqWalletService().getUnusedAddress().toString();
+            protocolModel.setBsqAddress(sellersBsqPayoutAddress);
 
             Coin buyersBsqChangeAmount = Coin.valueOf(tradePeer.getChange());
```
Would be safer to calculate this ourselves rather than taking from peer. Could verify that peer has the same value, but should not be necessary.

chimp1984 replied:
You refer to tradePeer.getChange? Will be handled in process msg task as discussed above.

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/tasks/buyer/ProcessBsqSwapFinalizeTxRequest.java:
```diff
@@ -39,20 +51,58 @@ public ProcessBsqSwapFinalizeTxRequest(TaskRunner<BsqSwapTrade> taskHandler, Bsq
+            // We will use only the sellers inputs from the tx so we do not verify anything else
+            byte[] tx = request.getTx();
+            Transaction sellersTransaction = protocolModel.getBtcWalletService().getTxFromSerializedTx(tx);
+            List<RawTransactionInput> rawInputs = request.getBtcInputs();
```
Would be good to verify that these are correct BTC inputs if a local bitcoin node is available.

chimp1984 replied:
I think that adds too much complexity. Worst case is the tx is invalid and the user need to do a spv resync.

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/tasks/buyer/ProcessBsqSwapFinalizeTxRequest.java:
```diff
@@ -39,20 +51,58 @@ public ProcessBsqSwapFinalizeTxRequest(TaskRunner<BsqSwapTrade> taskHandler, Bsq
+            checkArgument(change == 0 || Restrictions.isAboveDust(Coin.valueOf(change)),
+                    "BTC change must be 0 or above dust");
+
+            String sellersBsqPayoutAddress = request.getBsqPayoutAddress();
+            checkNotNull(sellersBsqPayoutAddress, "sellersBsqPayoutAddress must not be null");
+            checkArgument(!sellersBsqPayoutAddress.isEmpty(), "sellersBsqPayoutAddress must not be empty");
+
+            String sellersBtcChangeAddress = request.getBtcChangeAddress();
+            checkNotNull(sellersBtcChangeAddress, "sellersBtcChangeAddress must not be null");
+            checkArgument(!sellersBtcChangeAddress.isEmpty(), "sellersBtcChangeAddress must not be empty");
```
Verify valid addresses

chimp1984 replied:
See above discussion

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/tasks/buyer/ProcessBsqSwapFinalizeTxRequest.java:
```diff
@@ -39,20 +51,58 @@ public ProcessBsqSwapFinalizeTxRequest(TaskRunner<BsqSwapTrade> taskHandler, Bsq
+            checkArgument(sumInputs >= sellersBtcInputAmount,
+                    "Sellers BTC input amount do not match our calculated required BTC input amount");
+
+            long change = request.getBtcChange();
+            checkArgument(change == 0 || Restrictions.isAboveDust(Coin.valueOf(change)),
+                    "BTC change must be 0 or above dust");
```
Trusting change from peer, better to verify change using `BsqSwapCalculation`

chimp1984 replied:
Will add verification for that

chimp1984 committed:
Complete trade in case the tx broadcast got a timeout

chimp1984 committed:
Remove unused methods

chimp1984 committed:
Add scriptTypeId to RawTransactionInput

We will need that to know if its a segwit input or not.
Add constructors for converting a TransactionInput to
RawTransactionInput. Handle bitcoinSerialize according
to segwit flag.

This change should not break anything as we only add a
field to protobuf and RawTransactionInput is not used in
cases where we depend on the hash of the payload like in
offers.

chimp1984 committed:
Replace constructor call with fromProto

chimp1984 committed:
Remove unused methods

chimp1984 committed:
Use RawTransactionInput constructor instead of getRawInputFromTransactionInput

chimp1984 committed:
Move getRawInputFromTransactionInput to TradeWalletService

Don't change old behaviour for normal trade protocol methods
where we use segwit=false for serialisation.

chimp1984 committed:
Fix wrong exception, add more log details

chimp1984 committed:
Rename getTxSize to getVBytesSize

chimp1984 committed:
Fix check for segwit.
transactionInput.hasWitness would only work for signed
inputs but we need to do the fee calculation before the signing.

Adjust fee calculation values using vbytes
according to https://bitcoin.stackexchange.com/questions/87275/how-to-calculate-segwit-transaction-fee-in-bytes

chimp1984 committed:
Add handling for InsufficientMoneyException

This need to be tested further...

chimp1984 committed:
Add verification for peers change amount

chimp1984 committed:
Add verification for peers change

chimp1984 committed:
Store sellersBsqPayoutAmount in protocolModel.setPayout
at process input task and reuse that value in next task

chimp1984 committed:
Fix checks for change with trade fees and tx fees

chimp1984 committed:
Comment out balance check (temp for dev)

chimp1984 committed:
Do not handle dust here. We will handle it in caller

chimp1984 committed:
Fix handling for slightly overpaid BTC funds

If available funds is < dust it will add the dust change to miner fee
If the If available funds is > dust but the additional change output
would produce another dust change it will also add it to miner fees.
In that case the added amount is > dust.

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/tasks/buyer/BuyerCreatesBsqInputsAndChange.java:
```diff
@@ -66,9 +66,7 @@ protected void run() {
-        return BsqSwapCalculation.getVBytesSize(protocolModel.getTradeWalletService(),
-                Objects.requireNonNull(protocolModel.getInputs()),
-                protocolModel.getChange());
+        return BsqSwapCalculation.getVBytesSize(Objects.requireNonNull(protocolModel.getInputs()), protocolModel.getChange());
```
Prefer to keep lines within 120 characters.

chimp1984 replied:
OK

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/tasks/buyer/ProcessBsqSwapFinalizeTxRequest.java:
```diff
@@ -81,9 +82,20 @@ protected void run() {
+            }
+            if (DevEnv.isDevMode()) {
+                checkArgument(expectedChange == change);
+            }
```
Shouldn't this check always happen?

chimp1984 replied:
I removed that in a later commit as its an expected case when change is dust.

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/tasks/buyer/ProcessBsqSwapFinalizeTxRequest.java:
```diff
@@ -87,15 +87,13 @@ protected void run() {
             long expectedChange = sumInputs - buyersBtcPayout - sellersTxFee - buyersTxFee;
-            if (expectedChange != change) {
+            boolean isChangeAboveDust = Restrictions.isAboveDust(Coin.valueOf(expectedChange));
+            if (expectedChange != change && isChangeAboveDust) {
```
This check is a bit odd. If `expectedChange` is less than dust and different from `change` it won't trigger. There also is no `checkArgument(expectedChange == change)`

chimp1984 replied:
If input is x + dust and change would be dust we add the dust to miner fees. In that case the expectedChange and change are not the same and expectedChange < dust.
There are still other cases where input is x + minerfee for extra output + dust. here the change would first cause an additional output, thus triggering another fee calculation. then the change with the new output results in dust change and we give it to miners as well.

sqrrm replied:
The change value from peer still has to be checked that it's <= expectedChange. I don't think it's safe to not check it.

chimp1984 replied:
Yes agree.

sqrrm commented on core/src/main/java/bisq/core/trade/protocol/bsq_swap/tasks/seller/SellerCreatesAndSignsTx.java:
```diff
@@ -66,29 +63,32 @@ protected void run() {
+            Coin sellersBtcChangeAmount = Coin.ZERO;
             Coin required = BsqSwapCalculation.getSellersBtcInputValue(trade, sellersTxSize, sellersTradeFee);
 
+            // We do one iteration here to get the size of the inputs (segwit or not)
+            inputsAndChange = tradeWalletService.getSellersBtcInputsForBsqSwapTx(required);
+            sellersTxSize = BsqSwapCalculation.getVBytesSize(inputsAndChange.first, 0);
+            required = BsqSwapCalculation.getSellersBtcInputValue(trade, sellersTxSize, sellersTradeFee);
```
I don't understand why this is needed, seems it's covered in the loop

chimp1984 replied:
Yes thats another complicate issue. First iteration we use the min tx size of 104 vbytes. if you have the exact inputs for the expected tx where input is a non-segwit input it would create a change output (diff of larger tx size for non segwit and min tx size). that would trigger then trigger a calculation for 1 nonsegwit input (correct) + 1 output (not correct) and would cause a InsufficientMoneyException. To avoid that I set the change at the frist iteration to 0 so we get the correct size for the input and then do the loop where we get then the correct result with 0 change and the exact funds are used up.

Will likely require more tests with diff. such scenarios. But at least the ones I tested work with the current code.

chimp1984 committed:
Add address validation

chimp1984 committed:
Add check for change <= expectedChange

chimp1984 committed:
Update protocol description

chimp1984 committed:
Rename methods and params

chimp1984 committed:
Add getParentTxId method

Cleanup, make constructors private and
add deprecated annotation

chimp1984 committed:
Use getParentTxId method

chimp1984 committed:
Use isDust

chimp1984 committed:
Add getUnspentTxOutputValue method
Verify also sum of input values from DAO state data

chimp1984 committed:
Handle dust in caller
Add comments
